diff --git a/Original/src/afl-fuzz-run.c b/IgorFuzz/src/afl-fuzz-run.c
index 17c305e..5eff0b1 100755
--- a/Original/src/afl-fuzz-run.c
+++ b/IgorFuzz/src/afl-fuzz-run.c
@@ -352,7 +352,7 @@ u8 calibrate_case(afl_state_t *afl, struct queue_entry *q, u8 *use_mem,
   if (q->exec_cksum) {
 
     memcpy(afl->first_trace, afl->fsrv.trace_bits, afl->fsrv.map_size);
-    hnb = has_new_bits(afl, afl->virgin_bits);
+    hnb = has_few_bits(afl, afl->virgin_bits);    // update virgin_bits with the newly generated test case
     if (hnb > new_bits) { new_bits = hnb; }
 
   }
@@ -388,7 +388,7 @@ u8 calibrate_case(afl_state_t *afl, struct queue_entry *q, u8 *use_mem,
     cksum = hash64(afl->fsrv.trace_bits, afl->fsrv.map_size, HASH_CONST);
     if (q->exec_cksum != cksum) {
 
-      hnb = has_new_bits(afl, afl->virgin_bits);
+      hnb = has_few_bits(afl, afl->virgin_bits);      // initialize virgin_bits with input poc during first run!
       if (hnb > new_bits) { new_bits = hnb; }
 
       if (q->exec_cksum) {
